src/library/module_mgr.cpp

```diff
@@ -228,15 +234,16 @@ void module_mgr::build_lean(std::shared_ptr<module_info> const & mod, name_set c
     auto imports = get_direct_imports(id, mod->m_contents);
 
     mod->m_lt = logtree();
-    mod->m_trans_mtime = mod->m_mtime;
+    mod->m_trans_hash = mod->m_src_hash;
     for (auto & d : imports) {
         module_id d_id;
         std::shared_ptr<module_info const> d_mod;
         try {
             d_id = resolve(d, id);
             build_module(d_id, true, module_stack);
             d_mod = m_modules[d_id];
-            mod->m_trans_mtime = std::max(mod->m_trans_mtime, d_mod->m_trans_mtime);
+            // TODO(Vtec234): better mix
```

Comment:
I've thought about this a bit more, and I'm now inclined to say a simple xor is fine. There are no security concerns here so it can be predictable, and xoring *will* incorporate data from all imports. Can you think of a case where this breaks? The only one that comes to mind for me is if the same module is imported twice, then that would xor to 0 and `src_hash == trans_hash`. But a) Lean actually just ignores any duplicate import statements and b) even if it didn't, the `src_hash == trans_hash` would still be different for any two files that both happen to import some module twice.

Reply:
[This](https://github.com/leanprover-community/lean/blob/03d515a6f6556c9087a3e4abfb5245500766931e/src/library/module.cpp#L646) is the line is was basing that claim on. However, for the sake of not even having to make that argument, I added a different way to combine hashes.

Reply:
I think you looked in completely the wrong place.  The pipeline is roughly as follows: 1) the parser produces a list of names (e.g. `data.nat`, `tactic.core`, `data.nat`), 2) then the module_mgr resolves each of the names in the list to a file and recursively produces the module_info for each of the dependencies, 3) then the parser produces this list of names again, 4) then they are passed to `import_modules` which then uses a callback passed from the module_mgr to resolve the modules and imports them into the environment.

This function here is 2).  The `imports` list contains names, which may contain duplicates.  (There's some extra information in addition to the names for relative imports, but that's not so important.)  The line you've linked to is from 4), and that line just makes sure that we do not import the same module twice into an environment.

Reply:
Gotcha, thanks for explaining that! Maybe the parser could also include duplicate imports, but in any case since hashes are combined differently now, I am marking this as resolved.